File: docs/qa/qa2/run-repeat.py
import json, pathlib, subprocess, time
root = pathlib.Path(__file__).resolve().parents[3]
qa = root / 'docs/qa/qa2'
paths = subprocess.check_output(['git', 'ls-files', 'e2e/artifacts'], cwd=root, text=True).splitlines()
original = {p: (root / p).read_bytes() for p in paths}
started = time.monotonic()
try:
    with (qa / 'logs/e2e-repeat.log').open('w') as log:
        result = subprocess.run('npm --prefix e2e ci && npm --prefix e2e run test:repeat', shell=True, cwd=root, stdout=log, stderr=subprocess.STDOUT)
    for name in ['results.json', 'studio.png', 'question-1600.png', 'question-400.png']:
        source = root / 'e2e/artifacts' / name
        if source.exists():
            destination = qa / ('logs' if name.endswith('.json') else 'shots') / ('e2e-' + name)
            destination.write_bytes(source.read_bytes())
    (qa / 'logs/e2e-command.json').write_text(json.dumps({'exitCode':result.returncode,'elapsedSeconds':time.monotonic()-started}, indent=2))
finally:
    for p, data in original.items():
        (root / p).write_bytes(data)

File: docs/qa/qa3/run-repeat.mjs
import {readFile,writeFile,copyFile} from 'node:fs/promises';
import {execFileSync,spawn} from 'node:child_process';
const root=new URL('./',import.meta.url).pathname;
const paths=execFileSync('git',['ls-files','-z'],{encoding:'utf8'}).split('\0').filter(Boolean);
const before=new Map(await Promise.all(paths.map(async p=>[p,await readFile(p)])));
const started=new Date().toISOString(),start=performance.now();
const log=await import('node:fs').then(fs=>fs.createWriteStream(root+'logs/e2e-repeat.log'));
const child=spawn('/bin/zsh',['-c','npm --prefix e2e ci && npm --prefix e2e run test:repeat'],{stdio:['ignore','pipe','pipe']});child.stdout.pipe(log);child.stderr.pipe(log);
const code=await new Promise(resolve=>child.on('exit',resolve));
const changed=[];
for(const [p,original] of before){const now=await readFile(p);if(!now.equals(original)){const dest=root+(p.endsWith('.png')?'shots/e2e-':'logs/e2e-')+p.split('/').at(-1);await writeFile(dest,now);await writeFile(p,original);changed.push({path:p,copy:dest.slice(root.length)});}}
await writeFile(root+'logs/e2e-command.json',JSON.stringify({command:'npm --prefix e2e ci && npm --prefix e2e run test:repeat',started,finished:new Date().toISOString(),durationMs:performance.now()-start,exitCode:code,restoredTrackedFiles:changed},null,2));
console.log(JSON.stringify({exitCode:code,durationMs:performance.now()-start,restoredTrackedFiles:changed},null,2));process.exitCode=code;
